File: scm-java/src/main/scheme/primitives/PrimitiveThreadTerminateB.java
package scheme.primitives;
import scheme.*;

public class PrimitiveThreadTerminateB extends Primitive {
    @Override
    public String name() { return "thread-terminate!"; }

    @Override
    public String info() {
        return "Syntax: (thread-terminate! thread)\n" +
               "Library: (srfi 18)\n" +
               "Description: Terminates the given thread.\n" +
               "Example:\n" +
               "  (thread-terminate! thread)";
    }

    @Override
    public Object apply(SourcePos pos, Object[] arguments) {
        checkArgs(pos, arguments, 1, 1);
        SchemeThread t = (SchemeThread) Value.asNativeValue(arguments[0]).value;
        t.terminated = true;
        t.state = SchemeThread.State.TERMINATED;
        if (t == SchemeThread.currentThread.get()) {
            throw new SchemeError(pos, "thread terminated");
        }
        return Value.NIL;
    }
}

File: scm-csharp/scheme/primitives/PrimitiveThreadTerminateB.cs
namespace scheme;

public class PrimitiveThreadTerminateB : Primitive
{
    public override string Name() => "thread-terminate!";

    public override string Info() =>
        "Syntax: (thread-terminate! thread)\n" +
        "Library: (srfi 18)\n" +
        "Description: Terminates the given thread. If the target is the current thread, " +
        "an error is raised.\n" +
        "Example:\n" +
        "  (thread-terminate! t)";

    public override object Apply(SourcePos? pos, object[] arguments)
    {
        CheckArgs(pos, arguments, 1, 1);
        SchemeThread t = (SchemeThread) Value.AsNativeValue(arguments[0]).value;
        t.terminated = true;
        t.state = SchemeThreadState.TERMINATED;
        if (t == SchemeThread.CurrentThread)
            throw new SchemeError(pos, "thread terminated");
        return Value.NIL;
    }
}
